marshmallow_recipe/generics.py

```diff
@@ -0,0 +1,120 @@
+import dataclasses
+import types
+import typing
+from typing import TYPE_CHECKING, Annotated, Any, Generic, TypeAlias, TypeVar, Union, get_args, get_origin
+
+_GenericAlias: TypeAlias = typing._GenericAlias  # type: ignore
+
+if TYPE_CHECKING:
+    from _typeshed import DataclassInstance
+else:
+    DataclassInstance: TypeAlias = type
+
+TypeLike: TypeAlias = type | TypeVar | types.UnionType | types.GenericAlias | _GenericAlias
+FieldsTypeMap: TypeAlias = dict[str, TypeLike]
+TypeVarMap: TypeAlias = dict[TypeVar, TypeLike]
+FieldsClassMap: TypeAlias = dict[str, TypeLike]
+ClassTypeVarMap: TypeAlias = dict[TypeLike, TypeVarMap]
+FieldsTypeVarMap: TypeAlias = dict[str, TypeVarMap]
+
+
+def get_fields_type_map(cls: type) -> FieldsTypeMap:
+    origin: type = get_origin(cls) or cls
+    if not dataclasses.is_dataclass(origin):
+        raise ValueError(f"{origin} is not a dataclass")
+
+    class_type_var_map = get_class_type_var_map(cls)
+    fields_class_map = get_fields_class_map(origin)
+    return {
+        f.name: build_subscripted_type(f.type, class_type_var_map.get(fields_class_map[f.name], {}))
+        for f in dataclasses.fields(origin)
+    }
+
+
+def get_fields_class_map(cls: type[DataclassInstance]) -> FieldsClassMap:
+    names: dict[str, dataclasses.Field] = {}
+    result: FieldsClassMap = {}
+
+    mro = cls.__mro__
+    for cls in (*mro[-1:0:-1], cls):
```

Comment:
we are relying on an assumption how `dataclasses` resolves fields in hierarchy. This is a potentially fragile place

Reply:
`cls.__mro__[-1:0:-1]` - it is a copy from dataclasses implementation and unfortunately I have not found anything in dataclasses module which can solve my issue. that is why I copied it. we have tests and we will know if something is broken